R/weighted_efficacy.R

```diff
@@ -0,0 +1,275 @@
+#'
+#' Those eligible are those who have had a first dose more than a given period prior and
+#' who have not yet received a second dose.
+#'
+#' @param dose_times  Dose monitoring data
+#' @param t Current time
+#' @param dose_period  Days between dose 1 and dose 2
+eligable_for_second <- function(dose_times, t, dose_period){
+  lapply(dose_times, function(x){
+    had_first_beyond_threshold <- t - x[, 1] >= dose_period
+    had_first_beyond_threshold[is.na(had_first_beyond_threshold)] <- FALSE
+    not_had_2nd <- is.na(x[, 2])
+    had_first_beyond_threshold & not_had_2nd
+  })
+}
+
+#' Administer first doses
+#'
+#' @param ad Number of doses to administer in each age group
+#' @param dose_times Dose monitoring data
+#' @param t Current time
+administer_first_dose <- function(ad, dose_times, t){
+  for(i in seq_along(ad)){
+    if(ad[i] > 0){
+      to_jab <- which(is.na(dose_times[[i]][,1]))[1:ad[i]]
+      dose_times[[i]][to_jab,1] <- t
+    }
+  }
+  return(dose_times)
+}
+
+#' Administer second doses
+#'
+#' @param ad Number of doses to administer in each age group
+#' @param dose_times Dose monitoring data
+#' @param dose_period Days between dose 1 and dose 2
+#' @param t Current time
+administer_second_dose <- function(ad, dose_times, dose_period, t){
+  eligable <- eligable_for_second(dose_times, t, dose_period)
+  for(i in seq_along(ad)){
+    if(ad[i] > 0){
+      to_jab <- which(eligable[[i]])[1:ad[i]]
+      dose_times[[i]][to_jab,2] <- t
+    }
+  }
+  return(dose_times)
+}
+
+
+#' Extract dose numbers administers
+#'
+#' @param dose_times Dose monitoring data
+#' @inheritParams weighted_efficacy
+extract_dose_number <- function(dose_times, maxt){
+  output_1dose <- matrix(0, nrow = maxt, ncol = length(dose_times))
+  colnames(output_1dose) <- 1:length(dose_times)
+  output_2dose <- matrix(0, nrow = maxt, ncol = length(dose_times))
+  colnames(output_2dose) <- 1:length(dose_times)
+  for(t in 1:maxt){
+    for(a in 1:length(dose_times)){
+      output_1dose[t, a] <- sum(dose_times[[a]][,1] == t, na.rm = TRUE)
+      output_2dose[t, a] <- sum(dose_times[[a]][,2] == t, na.rm = TRUE)
+    }
+  }
+  output_1dose <- as.data.frame(output_1dose) %>%
+    dplyr::mutate(t = 1:maxt) %>%
+    tidyr::pivot_longer(-.data$t, names_to = "age_group", values_to = "n_1dose") %>%
+    dplyr::mutate(age_group = factor(.data$age_group, levels = 1:length(dose_times)))
+  output_2dose <- as.data.frame(output_2dose) %>%
+    dplyr::mutate(t = 1:maxt) %>%
+    tidyr::pivot_longer(-.data$t, names_to = "age_group", values_to = "n_2dose") %>%
+    dplyr::mutate(age_group = factor(.data$age_group, levels = 1:length(dose_times)))
+  output <- dplyr::left_join(output_1dose, output_2dose, by = c("t", "age_group"))
+  return(output)
+}
+
+
+#' Extract weighted efficacies
+#'
+#' @param dose_number  Dose number: 1 or 2
+#' @inheritParams weighted_efficacy
+add_weighted_efficacy <- function(dose_number, infection_efficacy, disease_efficacy, v1v2){
```

Comment:
Should there also be a dose 2 delay, i.e. currently this assumes that the efficacy conferred by the booster dose is realised immediately

Reply:
Yes, that's right - we need to check this, maybe with RTM?

Reply:
Have check. Suggest use a short delay of 7 days:

Mulligan MJ, Lyke KE, Kitchin N, Absalon J, Gurtman A, Lockhart S, et al. Phase I/II study of COVID-19 RNA vaccine BNT162b1 in adults. Nature. Nature Research; 2020;586: 589–593. 

Ramasamy MN, Minassian AM, Ewer KJ, Flaxman AL, Folegatti PM, Owens DR, et al. Safety and immunogenicity of ChAdOx1 nCoV-19 vaccine administered in a prime-boost regimen in young and old adults (COV002): a single-blind, randomised, controlled, phase 2/3 trial. Lancet. Lancet Publishing Group; 2020;396: 1979–1993.